crates/sui-data-ingestion-core/src/lib.rs

```diff
@@ -23,19 +24,19 @@ pub use worker_pool::WorkerPool;
 
 #[async_trait]
 pub trait Worker: Send + Sync {
-    async fn process_checkpoint(&self, checkpoint: &CheckpointData) -> Result<()>;
-    /// Optional method. Allows controlling when workflow progress is updated in the progress store.
-    /// For instance, some pipelines may benefit from aggregating checkpoints, thus skipping
-    /// the saving of updates for intermediate checkpoints.
-    /// The default implementation is to update the progress store for every processed checkpoint.
-    async fn save_progress(
-        &self,
-        sequence_number: CheckpointSequenceNumber,
-    ) -> Option<CheckpointSequenceNumber> {
-        Some(sequence_number)
-    }
+    type Result: Send + Sync;
+    async fn process_checkpoint(&self, checkpoint: &CheckpointData) -> Result<Self::Result>;
 
     fn preprocess_hook(&self, _: &CheckpointData) -> Result<()> {
         Ok(())
     }
 }
+
+#[async_trait]
+pub trait Reducer<R: Send + Sync>: Send {
+    async fn commit(&self, batch: Vec<R>) -> Result<CheckpointSequenceNumber>;
+    fn include_in_batch(&self, batch: &[R], element: &R) -> bool;
+    fn close_pending_batch(&self, _batch: &[R]) -> bool {
```

Comment:
would appreciate feedback on the interface. I tried to keep it minimalistic but couldn't get away with just two methods because different workflows may require different behavior.
For instance, the archival flow can't close a batch prematurely; it needs to accumulate a sufficient number of checkpoints, while the PG indexer prefers to commit the checkpoints it has accumulated so far.

An alternative would be to use `fn include_in_batch(&self, batch: &[R], element: Option<&R>) -> bool` to merge the two methods.
Or we could have a commit function that commits part of a batch and returns any unprocessed elements back to the loop. This was the original version, but Brandon pointed out that it could be error-prone for developers.

Reply:
I also thought over this, I think we can merge `include` and `close` into one
```
fn should_commit(&self, batch: &[R]) -> Option<Vec<R>>
```
together we will want to change the reducer codes tho, commenting there as well.

Reply:
The include_in_batch(&self, batch: &[R], element: &R) interface is necessary because you might want to exclude an element from a batch (for instance, if it's from a new epoch) and start a new batch instead. In this case, a signature like this is needed.

To simplify, close_pending_batch could be defined as close_pending_batch(&self) -> bool, returning true by default

Reply:
or, by following the idea listed below, we can have the following interface:
```
async fn commit(&self, batch: Vec<R>) -> Result<CheckpointSequenceNumber>;
async fn partial_commit(&self, batch: Vec<R>) -> Result<(CheckpointSequenceNumber, Vec<R>)> {
     let watermark = self.commit(batch).await?;
     Ok((warermark, vec![]))
}
```
So most of the devs will just need to implement the commit function, while more involved cases can override partial_commit